airbyte_cdk/sources/declarative/resolvers/config_components_resolver.py

```diff
@@ -91,17 +94,56 @@ def __post_init__(self, parameters: Mapping[str, Any]) -> None:
                 )
 
     @property
-    def _stream_config(self) -> Iterable[Mapping[str, Any]]:
-        path = [
-            node.eval(self.config) if not isinstance(node, str) else node
-            for node in self.stream_config.configs_pointer
-        ]
-        stream_config = dpath.get(dict(self.config), path, default=[])
+    def _stream_config(self) -> List[Dict[str, Any]]:
+        """
+        Builds a list of stream configuration dictionaries.
+
+        Each resulting config represents a unique combination of configurations
+        defined by the `configs_pointer` and any provided `default_values`.
+        """
+
+        def resolve_path(pointer: List[Union["InterpolatedString", str]]) -> List[str]:
+            """
+            Resolves a list of JSON pointer elements by evaluating interpolated strings
+            or using the raw strings directly.
+            """
+            return [
+                node.eval(self.config) if not isinstance(node, str) else node for node in pointer
+            ]
+
+        def prepare_streams() -> Iterable[List[Tuple[int, Any]]]:
+            """
+            Prepares all stream configs to an iterable where each item
+            is a list of (index, config) pairs for a particular stream.
+            """
+            for stream_config in self.stream_configs:
+                path = resolve_path(stream_config.configs_pointer)
+                stream_configs_raw = dpath.get(dict(self.config), path, default=[])
+                stream_configs = (
+                    list(stream_configs_raw)
+                    if isinstance(stream_configs_raw, list)
+                    else [stream_configs_raw]
+                )
+
+                if stream_config.default_values:
+                    stream_configs.extend(stream_config.default_values)
+
+                yield [(i, item) for i, item in enumerate(stream_configs)]
 
-        if not isinstance(stream_config, list):
-            stream_config = [stream_config]
+        def merge_combination(combo: Iterable[Tuple[int, Any]]) -> Dict[str, Any]:
+            """
+            Merges a combination of indexed config items into a single config dict.
+            """
+            result = {}
+            for config_index, (elem_index, elem) in enumerate(combo):
+                if isinstance(elem, dict):
+                    result.update(elem)
+                else:
+                    result.setdefault(f"source_config_{config_index}", (elem_index, elem))
+            return result
 
-        return stream_config
+        all_indexed_streams = list(prepare_streams())
```

Comment:
The result of `prepare_streams()` is an iterable, but we always iterate over it to turn it back into a list. We should probably not perform this `list()` conversion unless it must be turned into a list to process on the next line.

And if we were to do that, then we may want to just make `prepare_streams()` return a list for simplicity

Reply:
Remove prepare_streams() and move implementation to _stream_config